cirq/contrib/two_qubit_gates/math_utils.py

```diff
@@ -0,0 +1,292 @@
+
+    if sample_phase:
+        # coverage: ignore
+        global_phase = np.exp(1j * TWO_PI * np.random.rand(number))
+        np.einsum('t,tab->tab', global_phase, out, out=out)
+    return out
+
+
+def vector_kron(first: np.ndarray, second: np.ndarray) -> np.ndarray:
+    """Vectorized implementation of kron for square matrices."""
+    s_0, s_1 = first.shape[-2:], second.shape[-2:]
+    assert s_0[0] == s_0[1]
+    assert s_1[0] == s_1[1]
+    out = np.einsum('...ab,...cd->...acbd', first, second)
+    s_v = out.shape[:-4]
+    return out.reshape(s_v + (s_0[0] * s_1[0],) * 2)
+
+
+_MAGIC = np.array([[1, 0, 0, 1j], [0, 1j, 1, 0], [0, 1j, -1, 0], [1, 0, 0, -1j]
+                  ]) * np.sqrt(0.5)
+
+_gamma = np.array([[1, 1, 1, 1], [1, 1, -1, -1], [-1, 1, -1, 1], [1, -1, -1, 1]
+                  ]) * 0.25
+
+# all permutations of (1,2,3)
+_perms_123 = np.zeros((6, 3, 3), int)
+for ind, perm in enumerate(permutations((0, 1, 2))):
+    _perms_123[ind, (0, 1, 2), perm] = 1
+
+_negations = np.zeros((4, 3, 3), int)
+_negations[0, (0, 1, 2), (0, 1, 2)] = 1
+_negations[1, (0, 1, 2), (0, 1, 2)] = (1, -1, -1)
+_negations[2, (0, 1, 2), (0, 1, 2)] = (-1, 1, -1)
+_negations[3, (0, 1, 2), (0, 1, 2)] = (-1, -1, 1)
+
+_offsets = np.zeros((4, 3))
+_offsets[1, (1, 2)] = np.pi / 2
+_offsets[1, (0, 2)] = np.pi / 2
+_offsets[1, (0, 1)] = np.pi / 2
+
+
+def _kak_equivalent_vectors(kak_vec) -> np.ndarray:
+    """Generates all KAK vectors equivalent under single qubit unitaries."""
+    # coverage: ignore
+    kak_vec = np.asarray(kak_vec)
+    # Produce all shift-negations of the kak vector
+    out = np.einsum('nab,...b->...na', _negations, kak_vec,
+                    dtype=float)  # (...,4,3)
+    out[..., :, :] += _offsets
+
+    # Apply all permutations of indices
+    out = np.einsum('pcb,...nb->...pnc', _perms_123, out)  # (...,6,4,3)
+
+    # Merge indices
+    return np.reshape(out, out.shape[:-3] + (24, 3))
+
+
+def KAK_vector_infidelity(k_vec_a: np.ndarray,
+                          k_vec_b: np.ndarray,
+                          ignore_equivalent_vectors: bool = False
+                         ) -> np.ndarray:
+    """Minimum entanglement infidelity between two KAK vectors. """
+
+    if ignore_equivalent_vectors:
+        k_diff = k_vec_a - k_vec_b
+        out = 1 - np.product(np.cos(k_diff), axis=-1)**2
+        out -= np.product(np.sin(k_diff), axis=-1)**2
+        return out
+    # coverage: ignore
+    # We must take the minimum infidelity over all possible locally equivalent
+    # KAK vectors. We need only consider equivalent vectors of one input.
+    k_vec_a = np.asarray(k_vec_a)[..., np.newaxis, :]  # (...,1,3)
+    k_vec_b = _kak_equivalent_vectors(np.asarray(k_vec_b))  # (...,24,3)
+
+    k_diff = k_vec_a - k_vec_b
+
+    out = 1 - np.product(np.cos(k_diff), axis=-1)**2
+    out -= np.product(np.sin(k_diff), axis=-1)**2  # (...,24)
+
+    return out.min(axis=-1)
+
+
+def in_weyl_chamber(xp: np.ndarray, yp: np.ndarray,
```

Comment:
Why not take the same type that is returned by `kak_vector`, i.e. have the last axis be the length-3 x,y,z axis?

Reply:
Done.